Add barebones shop_admin LiveView

Baradoy opened the pull request:
Adds the most barebones shop admin liveview. Just enough to show that it renders inside of Shopify admin.

![image](https://github.com/orbit-apps/elixir-shopify-app/assets/16840642/2e39be6f-4e70-4a77-bbcb-511bab2a6e53)


https://github.com/orbit-apps/elixir-shopify-app/assets/16840642/c0d2cbb9-4201-438c-b848-3f2229c03e6f

Baradoy committed:
Add barebones shop_admin LiveView

Adds the most barebones shop admin liveview. Just enough to show that it renders inside of Shopify admin.

Baradoy commented on lib/shopify_app_web/endpoint.ex:
```diff
@@ -8,7 +8,8 @@ defmodule ShopifyAppWeb.Endpoint do
     key: "_shopify_app_key",
     signing_salt: "ijq1g2Yt",
-    same_site: "Lax"
+    same_site: "None",
```
I do not think that we want `same_site: None` for the entire endpoint. We will want to scope this to `/shop_admin/`, but it is not immediately obvious to me what that should look like.

Baradoy commented on lib/shopify_app_web/endpoint.ex:
```diff
@@ -45,10 +46,6 @@ defmodule ShopifyAppWeb.Endpoint do
 
-  if Mix.env() == :dev do
-    plug ShopifyApp.Plug.DevProxy, upstream: "http://localhost:3000"
-  end
```
We may need to re-add this when we are doing compatibility with react.

Baradoy committed:
Add shopify.loading to shop_admin LiveView

Adds using Shopify's loading bar. This requires splitting out the JS for shop admin into its own file.

Baradoy committed:
Add extra LiveViews to demonstrate navigation

Add shop_admin_live_view to Web to include ShopAdminComponents.

Baradoy committed:
Add title-bar items allowing

Demonstrate that we have access to the titlebar and titlebar actions

Baradoy committed:
Add toast messages when a message is put into flash

Baradoy commented on assets/js/hooks/index.js:
```diff
@@ -0,0 +1,15 @@
+
+See ShopAdminComponents.toast
+*/
+export const ShopifyToastHook = {
```
An alternate implementation that I am considering is to instead use `handleEvent` and handle a toast event. In my mind these are equally complicated but the current implementation seems closer to how flash usually works in LiveView.

Baradoy committed:
Change bindingPrefix for ShopAdmin to `data-phx-`

AppBridge is incredibly picky about what goes into things like `<ui-title-bar/>` and will error if `phx-` binding are there. Instead we can use `data-phx-`

Baradoy committed:
Add boilerplate readme

This replaces the ShopifyApp readme with something more relevant to your project. Started documenting the LiveView/AppBridge oddities there.

quinlanjager commented on lib/shopify_app_web/live/shop_admin/layouts/root.html.heex:
```diff
@@ -0,0 +1,29 @@
+    <%!-- Polaris CSS --%>
+    <link rel="stylesheet" href="https://unpkg.com/@shopify/polaris@11.14.0/build/esm/styles.css" />
+    <%!-- AppBridge JS --%>
+    <script src="https://cdn.shopify.com/shopifycloud/app-bridge.js">
+    </script>
```
Yo Graham, what's good? 😄

It seems UIs embedded into the Shopify environment are tricky to QA test with bots. I imagine this is because app views are usually tightly coupled to the AppBridge library (therefore running the UIs outside of Shopify is impossible).

My initial thought was that we could use an AppBridge "test harness" module that emulates some of its behaviour. That would let us run some UI tests in outside the Shopify environment.

If you're up for a bit of a "side-quest", I would like to get your thoughts on this idea and/or supporting automated UI testing in general, as you implement LiveView based Shopify admins.

Baradoy replied:
Hey Quinlan! Good to run into you again. 

There is already a very basic rendering test here, and LiveView gives us some decent [testing tools](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveViewTest.html). That will cover most of the cases that we could reasonably cover with a test harness. 

For Polaris, the best testing tools would be something like [Storybook](https://github.com/phenixdigital/phoenix_storybook). 

There still will have to be occasional QA inside of the Shopify Admin because Shopify can change AppBridge on us without notice. 

Thoughts on what a test harness would bring us that the existing testing tools and storybook do not already give us?

quinlanjager replied:
These testing tools land more on the development side of the product pipeline. Not to say that isn't good enough for most people. But I'm talking about automated in-browser tests, written with tools like Cypress. I see this as being valuable during QA, making tests easier to run, repeatable, and hopefully less mundane.

My current understanding is that app admins cannot run in a fully functional manner outside the Shop admin (due to the AppBridge deps) and Shopify makes it hard for bots to use their admins. So the test harness would allow us to boot up a Shop admin "danger room" to run these browser tests.

I should also note that I see this test harness as something the developer maintains (i.e. not part of this project), since maintaining a general purpose mock of the AppBridge is insane.

Baradoy committed:
Add AssignScope to allow authentication through the session

Instead of using a cookie for session storage, this just passes what is on the connecting conn into the session

Baradoy committed:
Use AssignScope for live_session validation

Baradoy commented on lib/shopify_app_web/endpoint.ex:
```diff
@@ -13,6 +13,14 @@ defmodule ShopifyAppWeb.Endpoint do
+  socket "/shop_admin_live", Phoenix.LiveView.Socket, websocket: [connect_info: []]
+  # This needs to be scoped to `/shop_admin` and have the orgins reduced.
+  # For some reason this is not working when put into a pipeline.
+  plug CORSPlug, origin: ["*"]
```
This should be fixed before being merged.
